runtime/src/qio/qio.c

```diff
@@ -2553,8 +2553,8 @@ qioerr _qio_buffered_behind(qio_channel_t* ch, int flushall)
           break;
         // no default to get warnings when new methods are added
       }
-      if( err ) goto error;
       qbuffer_iter_advance(&ch->buf, &write_start, num_written);
+      if( err ) goto error;
```

Comment:
The write call might have written some bytes and then had an error. That happens in some of the implementations of writev out of write (ie there is a loop calling write several times). If that happens, we've written 'num_written' and don't want to write it again, even though we are returning an error. Sometimes the error is transient and trying to write again will succeed. That is the case with EAGAIN, which we get with pipes sometimes.